dask/dataframe/core.py

```diff
@@ -383,19 +383,35 @@ def divisions(self):
 
     @divisions.setter
     def divisions(self, value):
+        if not isinstance(value, tuple):
+            raise TypeError("divisions must be a tuple")
+
+        if hasattr(self, "_divisions") and len(value) != len(self._divisions):
+            n = len(self._divisions)
+            raise ValueError(
+                f"This dataframe has npartitions={n - 1}, divisions should be a "
+                f"tuple of length={n}, got {len(value)}"
+            )
+
         if None in value:
             if any(v is not None for v in value):
-                warnings.warn(
-                    "recieved `divisions` with mix of nulls and non-nulls, future versions will only accept "
-                    "`divisions` that are all null or all non-null",
-                    PendingDeprecationWarning,
+                raise ValueError(
+                    "divisions may not contain a mix of None and non-None values"
                 )
-        if not isinstance(value, tuple):
-            warnings.warn(
-                f"recieved `divisions` of type {type(value)}, future versions will only accept `divisions` of type "
-                "tuple",
-                PendingDeprecationWarning,
-            )
+        else:
+            # Known divisions, check monotonically increasing
+
+            # XXX: if the index dtype is an ordered categorical dtype, then we skip the
+            # sortedness check, since the order is dtype dependent
+            index_dtype = getattr(self._meta, "index", self._meta).dtype
+            if not (is_categorical_dtype(index_dtype) and index_dtype.ordered):
+                if value != tuple(sorted(value)):
+                    raise ValueError("divisions must be sorted")
+            if len(value[:-1]) != len(list(unique(value[:-1]))):
+                raise ValueError(
+                    "divisions must be unique, except for the last element"
```

Comment:
I am a little hesitant about this assertion. It is entirely possible to have a well behaved dataframe with empty divisions. For instance if you do a slice. In that case the divisions that you get have dupes but they are still right.

Reply:
I'm not sure I follow how this would fail for empty divisions? Can you provide an example?

Reply:
Ok this is contrived:

```python
import dask

ddf = dask.datasets.timeseries(freq="15min") 
ddf = ddf.sort_values("name")
output  = ddf[ddf.name == "Zelda"]
```

The first 29 partitions are empty:

```python
[len(p) for p in output.partitions]
```

Reply:
I guess it's not really that it would fail for empty partitions and more what should the divisions be for empty partitions? Because right now we use duplicate values to indicate empty partitions. What can we use instead?

Reply:
Hmmm, that's a fair point, I hadn't thought of that. @gjoseph92 and I talked earlier today about this restriction and were pro adding it since it found and fixed a few bugs in the `set_index` code, but you're correct in that it doesn't work well for cases like this (as written). However, that code above doesn't trigger the error, since the divisions of `output` are still unknown, only when running the following is this check triggered:

```python
output.divisions = output.compute_current_divisions()
```

There is code in several places that assumes that `divisions` are all unique though (`loc` for example). We'll either want to harden those code bits to cases of duplicate divisions (doable), or provide a nicer workflow for culling duplicate divisions down (for example, the work done in `compute_current_divisions` would also be able to know which partitions are empty and could be dropped).

Reply:
Anyway, I think that for this specific PR dropping this check here is fine and makes sense. We should still keep the sorted checks (and I think the fixes done for the uniqueness check), but the actual assertion here is unnecessary.

Reply:
> I think there are perfectly valid reasons for having duplicate divisions

Very much agreed. I'd like to see us support this. But...

> I am also ok with leaving that check in since as you say there are places where non-empty divisions are assumed

Also agreed. The truth is that right now, we don't support duplicate divisions. The _current_ data model for DataFrame seems to assume that divisions are all unique. So adding logic to validate that that contract is upheld is a good thing, even if we want to change that contract down the line. IMO, making the current data model better-defined and better-validated will only make it easier to change that data model in the future, since you can more confidently know what the rest of the code is expecting, and how your changes need to change those expectations.

Reply:
Yeah I am happy with that agreement. Thanks for engaging with me on this.